src/auth/jwtclient.ts

```diff
@@ -155,57 +158,72 @@ export class JWT extends OAuth2Client {
    * @param {function=} callback Optional callback.
    */
-  fromJSON(json: JWTInput, callback?: (err?: Error) => void) {
-    const done = callback || noop;
-    if (!json) {
-      done(new Error(
-          'Must pass in a JSON object containing the service account auth settings.'));
-      return;
-    }
-    if (!json.client_email) {
-      done(new Error(
-          'The incoming JSON object does not contain a client_email field'));
-      return;
-    }
-    if (!json.private_key) {
-      done(new Error(
-          'The incoming JSON object does not contain a private_key field'));
-      return;
+  fromJSON(json: JWTInput, callback?: (err?: Error) => void): void {
+    try {
+      if (!json) {
+        throw new Error(
+            'Must pass in a JSON object containing the service account auth settings.');
+      }
+      if (!json.client_email) {
+        throw new Error(
+            'The incoming JSON object does not contain a client_email field');
+      }
+      if (!json.private_key) {
+        throw new Error(
+            'The incoming JSON object does not contain a private_key field');
+      }
+      // Extract the relevant information from the json key file.
+      this.email = json.client_email;
+      this.key = json.private_key;
+      this.projectId = json.project_id;
+    } catch (e) {
+      if (callback) {
+        callback(e);
+      } else {
+        throw e;
+      }
     }
-    // Extract the relevant information from the json key file.
-    this.email = json.client_email;
-    this.key = json.private_key;
-    this.projectId = json.project_id;
-    done();
+    if (callback) callback();
   }
 
   /**
    * Create a JWT credentials instance using the given input stream.
    * @param {object=} inputStream The input stream.
    * @param {function=} callback Optional callback.
    */
+  fromStream(inputStream: stream.Readable): Promise<void>;
   fromStream(
-      inputStream: stream.Readable, callback: (err?: Error|null) => void) {
-    const done = callback || noop;
-    if (!inputStream) {
-      setImmediate(() => {
-        done(new Error(
-            'Must pass in a stream containing the service account auth settings.'));
-      });
-      return;
+      inputStream: stream.Readable, callback: (err?: Error|null) => void): void;
+  fromStream(
+      inputStream: stream.Readable,
+      callback?: (err?: Error|null) => void): void|Promise<void> {
+    if (callback) {
+      this.fromStreamAsync(inputStream).then(r => callback()).catch(callback);
+    } else {
+      return this.fromStreamAsync(inputStream);
     }
-    let s = '';
-    inputStream.setEncoding('utf8');
-    inputStream.on('data', (chunk) => {
-      s += chunk;
-    });
-    inputStream.on('end', () => {
-      try {
-        const data = JSON.parse(s);
-        this.fromJSON(data, callback);
-      } catch (err) {
-        done(err);
+  }
+
+  private async fromStreamAsync(inputStream: stream.Readable) {
```

Comment:
nit: drop `async` unless you're awaiting.

Reply:
fixed